### PR TITLE
Add submission tags

### LINKED ISSUE #1347: Allow tagging submissions
It would be really useful to tag submissions and a view where we can find submissions with tags.

* Misgraded exercises
* suspicious submissions
* etc.

### PR DESCRIPTION
**Pull request by mikaelGusse**
# Description

**What?**

Added tags to submissions and the possibility to filter using these tags.

https://github.com/apluslms/a-plus/assets/159019283/adc4545d-72d3-4c62-90bb-817a303c28f5

**Why?**

Adds a requested feature that allows teachers to more easily identify particular submissions.

**How?**

Added a new django model for the submissiontag and additionally a submissiontagging class which stores which submissiontags belong to which submissions. Additionally added javascript table manipulation for filtering.

Fixes #1347

# Testing

**Remember to add or update unit tests for new features and changes.**

* How to [test your changes in A-plus](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations)
* How to [test accessibility](https://wiki.aalto.fi/display/EDIT/How+to+check+the+accessibility+of+pull+requests)

**What type of test did you run?**

- [ ] Accessibility test using the [WAVE](https://wave.webaim.org/extension/) extension.
- [x] Django unit tests.
- [x] Selenium tests.
- [ ] Other test. *(Add a description below)*
- [x] Manual testing.

Tested adding tags to multiple submissions and creating new tags. Also tried filtering with different combinations of tags.

**Did you test the changes in**

- [ ] Chrome
- [x] Firefox
- [ ] This pull request cannot be tested in the browser.

**Think of what is affected by these changes and could become broken**

# Translation

- [x] Did you modify or add new strings in the user interface? ([Read about how to create translation](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations))

# Programming style

- [x] Did you follow our [style guides](https://apluslms.github.io/contribute/styleguides/)?
- [x] Did you use Python type hinting in all functions that you added or edited? ([type hints](https://docs.python.org/3/library/typing.html) for function parameters and return values)

# Have you updated the README or other relevant documentation?

- [ ] documents inside the doc directory.
- [ ] README.md.
- [ ] Aplus Manual.
- [ ] Other documentation (mention below which documentation).

# Is it Done?

- [ ] Reviewer has finished the code review
- [ ] After the review, the developer has made changes accordingly
- [ ] Customer/Teacher has accepted the implementation of the feature

*Clean up your git commit history before submitting the pull request!*

### CODE REVIEW

File: exercise/templates/exercise/staff/_submissions_table.html [hunk 1]
```diff
@@ -124,3 +140,54 @@
 </table>
 
 {% include "exercise/staff/_regrade_submissions.html" %}
+<script>
```
Comment (murhum1) [inline exercise/templates/exercise/staff/_submissions_table.html:143]: This JS can definitely be moved to its own file since there's no django rendering involved. So move this to `exercise/static/exercise/filter_submissions_by_tag.js` or such, and then load here like 

```
<script src="{% static 'exercise/filter_submissions_by_tag.js' %}"></script>
```

File: exercise/templates/exercise/staff/list_submissions.html
Comment (murhum1) [inline exercise/templates/exercise/staff/list_submissions.html:1]: Unnecessary whitespace addition here

File: edit_course/templates/edit_course/submissiontag_list.html [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+				</a>
+				<a href="{{ tag|url:'course-submission-tags-remove' }}" class="aplus-button--secondary aplus-button--xs" role="button">
+					<span class="glyphicon glyphicon-remove" aria-hidden="true"></span>
+					{% translate "REMOVE" %}
+				</a>
+			</td>
+		</tr>
+		{% empty %}
+		<tr>
+			<td colspan="4">{% translate "NO_SUBMISSION_TAGS" %}</td>
```
Comment (murhum1) [inline edit_course/templates/edit_course/submissiontag_list.html:52]: There's no translation for this yet in english or finnish.

File: locale/fi/LC_MESSAGES/django.po [hunk 3]
```diff
@@ -2842,6 +2858,14 @@ msgstr "Poista käyttäjämerkintä"
 msgid "REMOVE_USER_TAG_CONFIRMATION"
 msgstr "Vahvista käyttäjämerkinnän poisto"
```
Comment (murhum1) [inline locale/fi/LC_MESSAGES/django.po:2860]: The finnish translation file is missing `REMOVE_SUBMISSION_TAG`, making it fall back to english

image

**Review by murhum1 — CHANGES_REQUESTED** [review]
Thanks! I left comments on some things I think should be fixed before merging. 

There's also some parts I'm not entirely sure about, and the whole thing with saving submission taggings by creating on-demand forms in JS, but we can take a look at that later if need be since this implementation does work.

File: edit_course/templates/edit_course/submissiontagging_add.html
[the remark this answers is not in the document]
Comment (mikaelGusse) [reply edit_course/templates/edit_course/submissiontagging_add.html:1]: I use the CreateView for creating these SubmissionTaggings so I need this so that I can use this form to create them. The user cannot normally navigate to this but they can access it using the url. Is that ok or should i do something else for adding the taggings?

File: exercise/templates/exercise/staff/_assessment_panel.html [hunk 4]
```diff
@@ -22,6 +24,31 @@
...
+				</div>
+				{% endfor %}
+				</div>
+				</p>
+
+				<div class="dropdown">
+				  <button class="btn btn-default dropdown-toggle" type="button" id="dropdownMenuButton" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+					{% translate 'ADD_TAGGING' %}
+				  </button>
+				  <ul class="dropdown-menu" aria-labelledby="dropdownMenuButton" data-tagid="subtag.id">
```
[the remark this answers is not in the document]
Comment (mikaelGusse) [reply exercise/templates/exercise/staff/_assessment_panel.html:45]: data-tagid was a vestigial attribute used previously for testing. Removed

File: exercise/templates/exercise/staff/_assessment_panel.html [hunk 5]
```diff
@@ -184,3 +211,56 @@
...
+	margin: 0.25em 0.75em 0.75em;
+	height: 15px;
+}
+.dropdown-menu {
+	position: absolute;
+	top: 35px;
+}
+</style>
+
+<script>
```
[the remark this answers is not in the document]
Comment (mikaelGusse) [reply exercise/templates/exercise/staff/_assessment_panel.html:217]: Yes. I think including the submission tags in the api and thus adjusting their state through that would be a good idea for the future.

File: exercise/templates/exercise/staff/_assessment_panel.html
[the remark this answers is not in the document]
Comment (mikaelGusse) [reply exercise/templates/exercise/staff/_assessment_panel.html:1]: Fixed this

File: exercise/templates/exercise/staff/list_submissions.html
Comment (mikaelGusse) [reply exercise/templates/exercise/staff/list_submissions.html:1]: Removed

File: locale/fi/LC_MESSAGES/django.po (hunk 3 above)
Comment (mikaelGusse) [reply locale/fi/LC_MESSAGES/django.po:2864]: Added

File: exercise/templates/exercise/staff/_assessment_panel.html [hunk 6]
```diff
@@ -22,6 +24,31 @@
...
+					{{tag.tag|colortag}}
+					<button type="submit" class="aplus-button--danger aplus-button--secondary aplus-button--xs"><span class="glyphicon glyphicon-remove" aria-hidden="true"></span></button>
+				  </form>
+				</div>
+				{% endfor %}
+				</div>
+				</p>
+
+				<div class="dropdown">
+				  <button class="btn btn-default dropdown-toggle" type="button" id="dropdownMenuButton" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
```
[the remark this answers is not in the document]
Comment (mikaelGusse) [reply exercise/templates/exercise/staff/_assessment_panel.html]: Added

File: exercise/templates/exercise/staff/_assessment_panel.html [hunk 7]
```diff
@@ -22,6 +24,31 @@
...
+					{% translate 'ADD_TAGGING' %}
+				  </button>
+				  <ul class="dropdown-menu" aria-labelledby="dropdownMenuButton" data-tagid="subtag.id">
+					{% for subtag in instance.submissiontags.all %}
+							<li><a class="dropdown-item" id={{subtag.id}} href="#" onclick="submitForm('{% url 'add-tag-to-submissions' course_slug=course.url instance_slug=instance.url module_slug=module.url exercise_path=exercise.get_path submission_id=submission.id subtag_id=subtag.id %}')"><span>{{subtag|colortag}}</span></a></li>
```
[the remark this answers is not in the document]
Comment (mikaelGusse) [reply exercise/templates/exercise/staff/_assessment_panel.html]: Added

File: exercise/templates/exercise/staff/_assessment_panel.html [hunk 8]
```diff
@@ -184,3 +211,56 @@
 		</div>
 	</div>
 </div>
+
+<input type="hidden" name="csrfmiddlewaretoken" value="{{ csrf_token }}">
```
[the remark this answers is not in the document]
Comment (mikaelGusse) [reply exercise/templates/exercise/staff/_assessment_panel.html]: Removed

File: exercise/templates/exercise/staff/_assessment_panel.html [hunk 9]
```diff
@@ -184,3 +211,56 @@
...
+
+<style>
```
[the remark this answers is not in the document]
Comment (mikaelGusse) [reply exercise/templates/exercise/staff/_assessment_panel.html]: Renamed

File: locale/en/LC_MESSAGES/django.po [hunk 10]
```diff
@@ -2827,6 +2844,19 @@ msgstr "Remove user tag"
...
+msgid "REMOVE_SUBMISSION_TAG_CONFIRMATION"
+msgstr "Remove submission tag"
+
+#: edit_course/templates/edit_course/submissiontag_delete.html
+msgid "REMOVE_SUBMISSION_TAG"
+msgstr "Remove submission tag"
+
+#: edit_course/templates/edit_course/submissiontag_delete.html
+msgid "REMOVE_SUBMISSION_TAG_CONFIRMATION_ALERT"
+msgstr "Confirm submission tag removal"
```
[the remark this answers is not in the document]
Comment (mikaelGusse) [reply locale/en/LC_MESSAGES/django.po]: Adjusted

File: locale/fi/LC_MESSAGES/django.po [hunk 11]
```diff
@@ -2842,6 +2858,14 @@ msgstr "Poista käyttäjämerkintä"
...
 
+#: edit_course/templates/edit_course/submissiontag_delete.html
+msgid "REMOVE_SUBMISSION_TAG_CONFIRMATION"
+msgstr "Poista palautusmerkintä"
+
+#: edit_course/templates/edit_course/submissiontag_delete.html
+msgid "REMOVE_SUBMISSION_TAG_CONFIRMATION_ALERT"
+msgstr "Vahvista palautusmerkinnän poisto"
```
[the remark this answers is not in the document]
Comment (mikaelGusse) [reply locale/fi/LC_MESSAGES/django.po]: Adjusted

File: exercise/templates/exercise/staff/_submissions_table.html [hunk 12]
```diff
@@ -79,7 +90,7 @@
 		</thead>
 		<tbody>
 				{% for submission in submissions %}
-				<tr>
+				<tr id="{{ submission.id }}">
```
[the remark this answers is not in the document]
Comment (mikaelGusse) [reply exercise/templates/exercise/staff/_submissions_table.html]: Added

File: exercise/templates/exercise/staff/_submissions_table.html (hunk 1 above)
Comment (mikaelGusse) [reply exercise/templates/exercise/staff/_submissions_table.html]: Moved to own file

File: edit_course/templates/edit_course/submissiontag_list.html (hunk 2 above)
Comment (mikaelGusse) [reply edit_course/templates/edit_course/submissiontag_list.html:52]: Added translations

File: exercise/templates/exercise/staff/list_submissions.html
Comment (murhum1) [reply exercise/templates/exercise/staff/list_submissions.html:1]: I was more referring to the fact that the only changes in the file in the commit were the two empty lines before `{% endblock %}, not the EOF whitespace (which is fine and apparently even recommended so some older tools don't misbehave), should've been clearer in my comment

File: edit_course/templates/edit_course/submissiontagging_add.html
[the remark this answers is not in the document]
Comment (murhum1) [reply edit_course/templates/edit_course/submissiontagging_add.html:1]: I took the liberty of handling this by removing the dependency on the form, and then removed some now-unnecessary things

**Review by murhum1 — APPROVED** [review]
